File: 125-valid-palindrome/valid-palindrome.java
class Solution {
    public boolean isPalindrome(String s) {
        HashSet<Character> alpNum = new HashSet<>();
        for(char c='a'; c<='z'; c++){
            alpNum.add(c);
        }
        for(char c='A'; c<='Z'; c++){
            alpNum.add(c);
        }
        for(char c='0'; c<='9'; c++){
            alpNum.add(c);
        }
        return checkPalindrome(s, alpNum);
    }
    public boolean checkPalindrome(String str, HashSet<Character> alpNum){
        int l=0;
        int r=str.length()-1;
        while(l<r){
            while(l<str.length() && l<r && !alpNum.contains(str.charAt(l))){
                l++;
            }
            if(l==str.length() || l>=r){
                break;
            }
            char ch_l = str.charAt(l);
            if(ch_l>='A' && ch_l<='Z'){
                ch_l = (char) (ch_l - 'A' + 'a');
            }

            while(r>=0 && l<r && !alpNum.contains(str.charAt(r))){
                r--;
            }
            if(r<0 || l>=r){
                break;
            }
            char ch_r = str.charAt(r);
            if(ch_r>='A' && ch_r<='Z'){
                ch_r = (char) (ch_r - 'A' + 'a');
            }        
            if(ch_l != ch_r){
                return false;
            }
            l++;
            r--;
        }
        return true;
    }
}

File: 125-valid-palindrome/valid-palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        l=0
        r=len(s)-1
        while l<r:
            while l<r and not ((s[l]>='a' and s[l]<='z') or (s[l]>='A' and s[l]<='Z') or (s[l]>='0' and s[l]<='9')):
                l += 1
            while l<r and not ((s[r]>='a' and s[r]<='z') or (s[r]>='A' and s[r]<='Z') or (s[r]>='0' and s[r]<='9')):
                r -= 1
            if s[l].lower() != s[r].lower():
                return False
            l += 1
            r -= 1
        return True